### ISSUE TITLE
Add tutorial on getting accessToken and shop name and creating simple REST API call

### ISSUE THREAD

**Issue by SerhiDi** [Shopify/shopify-app-template-node#621]
In the current app version no information about how we can store accessToken, shop name, and use they.

_server/server.js_
```
!!! CustomSessionStorage !!!

import RedisStore from './redis-store';
const sessionStorage = new RedisStore();

Shopify.Context.initialize({
  API_KEY: process.env.SHOPIFY_API_KEY,
  API_SECRET_KEY: process.env.SHOPIFY_API_SECRET,
  SCOPES: process.env.SCOPES.split(','),
  HOST_NAME: process.env.HOST.replace(/https:\/\//, ''),
  API_VERSION: ApiVersion.January21,
  IS_EMBEDDED_APP: true,
  SESSION_STORAGE: new Shopify.Session.CustomSessionStorage(
    sessionStorage.storeCallback,
    sessionStorage.loadCallback,
    sessionStorage.deleteCallback,
  )
});

------
!!! accessToken !!!

 createShopifyAuth({
  async afterAuth(ctx) {
  // Access token and shop available in ctx.state.shopify
  const {shop, accessToken, scope} = ctx.state.shopify;
```

For example, get themes (using _isomorphic-fetch_ and _koa-router_):

_server/routes.js_
```
const baseURL =  `https://${SHOPIFY_API_KEY}:${accessToken}@${shopOrigin}/admin/api/${ApiVersion.January21}`;

router.get('/themeId', async (ctx) => {

  const res = await fetch(
    `${baseUrl}/themes.json`
  );

  try {
      ctx.body = await res.json();
      ctx.status = 200;
  } catch (e) {
    console.log('error:', e.message);
  }
});
```

SHOPIFY_API_KEY - is environment variable, OK
But how I can save **shop** and **accessToken**?

The official tutorial doesn't contain any information about this.

**Comment by SerhiDi** [comment]
@paulomarg,

I added this code:
```
const session = await Shopify.Utils.loadCurrentSession(ctx.request, ctx.response, false);
console.log('SESSION', session);

const shopOrigin = session.shop;
const accessToken = session.accessToken;

const baseURL =  `https://${SHOPIFY_API_KEY}:${accessToken}@${shopOrigin}/admin/api/${ApiVersion.January21}`;

router.get('/themeId', async (ctx) => {

  const res = await fetch(
    `${baseUrl}/themes.json`
  );

  try {
      ctx.body = await res.json();
      ctx.status = 200;
  } catch (e) {
    console.log('error:', e.message);
  }
});
```

When I reload the app server and reload the app page in the browser I see:
- when app loading (is reinstalling after server reload) 
I see in log 'SESSION' - session object, all OK.

- after loading, when the app page opened,
I see in log 'SESSION' - _undefined_

How it works?
Why is there no example with _Shopify.Utils.loadCurrentSession_ in the application code?

**Comment by SerhiDi** [comment]
@paulomarg,
You have at least two developers who have tested the code from the tutorial and who have the same errors.
Сould you please check the code you referenced? [that repo's documentation](https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/rest.md)

**Comment by SerhiDi** [comment]
It did not shed light on why `Shopify.Utils.loadCurrentSession(ctx.request, ctx.response, false);` still `undefined`

**Comment by paulomarg** [comment]
Hi @SerhiDi, thank you for your feedback. We will take it into consideration to continue to improve our documentation. We do provide working code examples like the one I mentioned, which you can add to any action in your app as long as the request is made using `authenticatedFetch` in your client.

However, repeatedly mentioning me or any other team member will not expedite a response, so we ask that you please don't do that. GitHub issues are meant to track issues with our code, and we strive to address any actual problems as quickly as we can. Unfortunately, we don't have the bandwidth to support every partner individually, but we encourage partners to discuss their apps and solutions in the [Community Forums](https://community.shopify.com/c/App-Partner-Platform/ct-p/appdev), and report back to us via GitHub issues if there is evidence that our code is doing something incorrectly.

In the future, please make sure you adhere to the [Code of Conduct](https://github.com/Shopify/shopify-app-node/blob/master/.github/CODE_OF_CONDUCT.md) when interacting in our repositories.

**Comment by SerhiDi** [comment]
I created a new app,
https://github.com/SerhiDi/session-bug

The only change is these lines:
https://github.com/SerhiDi/session-bug/blob/3f8114cf8476e5fa1b2ec07e45917feef5eb1a02/server/server.js#L82-L86

If you try open app route link:
https://my-store.myshopify.com/admin/apps/app-name/token
you will get `undefined`.

I can state the fact that the function `Shopify.Utils.loadCurrentSession` does not work.
And the question remains open, how to get a token inside the route.

Also don't see an example with `authenticatedFetch`. I can get `session token` there, but not `accessToken`

**Comment by jennyl** [comment]
Hello. Providing the JWT in the header resolved the undefined issue for me when trying to make a REST API call from the front end.

```js
// pages/index.js
import { useAppBridge } from "@shopify/app-bridge-react";
import { getSessionToken } from "@shopify/app-bridge-utils";

function Index() {
   const app = useAppBridge();

   async function getCheckouts() {
       const token = await getSessionToken(app);

       const response = await fetch("/api/checkouts", {
           headers: { "Authorization": `Bearer ${token}` }
       });
       const checkouts = await response.json();

       console.log(checkouts);
   }

  return (
    <Page>
    </Page>
  );
}


// server/server.js
  router.get("/api/:object", verifyRequest({ returnHeader: true }), async (ctx) => {
    // Load the current session to get the `accessToken`.
    const session = await Shopify.Utils.loadCurrentSession(ctx.req, ctx.res);

    // Create a new client for the specified shop.
    const client = new Shopify.Clients.Rest(session.shop, session.accessToken);

    // Use `client.get` to request the specified Shopify REST API endpoint, e.g. `products`.
    const results = await client.get({
      path: ctx.params.object,
    });

    ctx.body = results.body;
    ctx.res.status = 200;
  });
```

Hope this helps anyone who is still struggling.

**Comment by Goose99** [comment]
> Hello. Providing the JWT in the header resolved the undefined issue for me when trying to make a REST API call from the front end.
> 
> ```js
> // pages/index.js
> import { useAppBridge } from "@shopify/app-bridge-react";
> import { getSessionToken, authenticatedFetch } from "@shopify/app-bridge-utils";
> 
> function Index() {
>    const app = useAppBridge();
>    //const aFetch = authenticatedFetch(app);
> 
>    async function getCheckouts() {
>        const token = await getSessionToken(app);
> 
>        const response = await fetch("/api/checkouts", {
>            headers: { "Authorization": `Bearer ${token}` }
>        });
>        //const response = await aFetch("/api/checkouts");
>        const checkouts = await response.json();
> 
>        console.log(checkouts);
>    }
> 
>   return (
>     <Page>
>     </Page>
>   );
> }
> ```
> 
> ```js
> // server/server.js
>   router.get("/api/:object", verifyRequest({ returnHeader: true }), async (ctx) => {
>     // Load the current session to get the `accessToken`.
>     const session = await Shopify.Utils.loadCurrentSession(ctx.req, ctx.res);
> 
>     // Create a new client for the specified shop.
>     const client = new Shopify.Clients.Rest(session.shop, session.accessToken);
> 
>     // Use `client.get` to request the specified Shopify REST API endpoint, e.g. `products`.
>     const results = await client.get({
>       path: ctx.params.object,
>     });
> 
>     ctx.body = results.body;
>     ctx.res.status = 200;
>   });
> ```
> 
> Hope this helps anyone who is still struggling.

Yes it helps but still means you cannot run API calls in getServerSideProps nextjs as getSessionToken wont work as not in app bridge

**Comment by SerhiDi** [comment]
Congratulations to all involved, nothing has appeared in the documentation for a year and a half.

**Comment by cquemin** [comment]
We have made quite a few changes in the v6 of [shopify-api](https://github.com/Shopify/shopify-api-js). In particular session storage implementations (and the interface they all implement) are now [in their own packages](https://github.com/Shopify/shopify-app-js/tree/main/packages). This should give greater flexibility for developers to implement their own session storage should they want to. Or they can use one of many implementation that we provide there.
I have run through the last version of the tutorial on [shopify.dev](https://shopify.dev/apps/getting-started/create) and everything works as expected.
Closing this one for now.
